### ISSUE TITLE
OO Problem class & method compatibility checking

### ISSUE THREAD

**Issue by ConnectedSystems** [SALib/SALib#234]
This will be a long read but I hope to spark some discussion around improvements to SALib in terms of usability in the longer term.

Before I start I'd like to thank everyone involved in making this package, it certainly has made doing this kind of work more accessible.

In using this package there are some pain points I have encountered which I'd like to show examples of, and suggest possible solutions. Some of the issues I raise below are relatively minor, but addressing these could improve usability.

Starting with the most minor:

1. Redundant attribute in "problem" specification

```python
problem = {
  'num_vars': 3,
  'names': ['x1', 'x2', 'x3'],
  'bounds': [[-np.pi, np.pi]]*3
}
```

I've always thought "why do I have to specify the number of parameters? It could be inferred from the number of names."

But I've noticed that some methods directly modify this value (in ff.py for example). This leads into the next suggestion of reconceptualizing how SALib represents the different sampling/analysis methods. Currently they are a set of functions which were developed by different contributors - and nothing wrong with that! But it does mean that methods are not always compatible (more on this later).

2. Objects for SALib

Previously it was suggested that a move towards a more object-oriented approach be made [as in here](https://github.com/SALib/SALib/issues/216#issuecomment-435647632). I previously pushed for keeping things as they currently are for the short-term but given the context here is for the medium/long term changes it's worth revisiting.

To implement my first suggestion would 1) require all methods to have the check implemented and create the `num_vars` element if it doesn't exist, 2) be reorganized into a class hierarchy from which the check is inherited, or 3) use a common decorator function that implements the check.

The first would introduce unnecessary duplication and complexity so that's off the table. The second is viable but would mean a significant overhaul of how things are implemented, with implications for the testing setup that is already in place. The third approach is what I'm currently leaning towards as it retains most of the current implementation (see further below for an example).

3. Mixing and matching methods

Currently users are able to mix and match different sampling and analysis methods without any indication that it is sane to do so. This leads to some confusion [as in here](https://github.com/SALib/SALib/issues/218) and [here](https://github.com/SALib/SALib/issues/224); addressing this pain point would go a long way to making SALib more user-friendly.

4. An example

To bring all the points above together, I'll adapt the canonical example from the readme.

I should also ask at this point if there's been any previous discussion around what the envisioned OO approach looks like.

The below is only an idea, none of the below has been tested, implemented, or checked for sanity, but I think it is doable.

```python
# This, with or without `num_vars`, would be a valid spec
problem = {
  'names': ['x1', 'x2', 'x3'],
  'bounds': [[-np.pi, np.pi]]*3
}

# Could also define like this:
problem = ProblemSpec(
  names=['x1', 'x2', 'x3'],
  bounds=[[-np.pi, np.pi]]*3
)

# In the process of generating samples, the ProblemSpec
# object would be updated with the sampling method used,
# the sample values, and a list of compatible methods.
param_values = latin.sample(problem, 1000)

# Run model
Y = Ishigami.evaluate(param_values)

# Perform analysis (incompatible method)
Si = sobol.analyze(problem, Y, print_to_console=True)
# The above would raise an error with a message like
# Error: SALib.sample.latin samples are not compatible with SALib.analyze.sobol
#        Compatible methods are ...

# Alternatively, in the case where the user specifies their own sampling values
# could continue on as normal but with a warning about not being able to check compatibility
```

A more OO approach (with method chaining) would be:

```python
problem = ProblemSpec(
  names=['x1', 'x2', 'x3'],
  bounds=[[-np.pi, np.pi]]*3
)

Si = problem\
      .sample('latin', num_samples=1000)\
      .evaluate(Ishigami.evaluate)\
      .analyze('sobol', print_to_console=True)
```

Both approaches could be supported without breaking backwards compatibility.

The question then is how to implement the above behaviour without significant rework. The most elegant solution (read: least amount of work) could be to have an implementation of the `ProblemSpec` (or simply `Problem`?) class above extend the base `dict` object, similar to the [previously implemented `ResultDict`](https://github.com/SALib/SALib/blob/master/src/SALib/util/results.py) to achieve what we want:

```python
# In util/ProblemSpec.py
import numpy as np


class ProblemSpec(dict):
  # Note this class inherits from the base Dictionary class so
  # its default behaviour is like a dict

  # A class property mapping samplers to compatible analysis functions
  # A more modular approach would be for each sampler to specify what is compatible
  compat_map = {
    'SALib.latin.sample': []
  }

  # Could dynamically load in available sampling/analysis methods here too


  def __init__(self, *args, **kwargs):
    super(ProblemSpec, self).__init__(*args, **kwargs)
    self.sampler_name = None

  def is_compatible(self, analyzer_name):
    compat_methods = ProblemSpec.compat_map[self.sampler_name]
    if not analyzer_name in compat_methods:
      err_msg = "Error: {} samples are not compatible with {}\n".format(self.sampler_name, analyzer_name)
      err_msg += "Compatible methods are:\n{}".format(compat_methods)
      raise TypeError(err_msg)
      # Alternatively, could simply return False and incompatibility error elsewhere...
    
    return True
  
  # to support the OO approach...
  def sample(self, sampler, num_samples, seed=None):
    # call the relevant samplers and store values
    self.samples = ...
    return self

  def evaluate(self, func):
    self.model_results = func(self.samples)
    return self

  def analyze(self, analyzer):
    self.analysis_results = ...
    return ResultDict(...)

# End ProblemSpec()
```

The ProblemSpec class could then be used in conjunction with decorators. The advantage of this approach is that backwards compatibility could be maintained without a significant rewrite, and future contributors would simply decorate their submitted methods.

```python
# In util/method_decorators.py
from functools import wraps
from types import MethodType
from .results import ResultDict


def sampler_decorator(func):
  @wraps(func)
  def sampler_wrapper(*args, **kwargs):
    if 'num_vars' not in problem:
      problem['num_vars'] = len(problem['names'])

    if not issubclass(problem, ProblemSpec):
      # Convert dict to ProblemSpec if necessary
      problem = ProblemSpec(**problem)

    samples = func(*args, **kwargs)
    sampler_name = func.__module__

    # Compatible methods could be specified here to make it more modular.

    problem.samples = samples
    problem.sampler_name = sampler_name
    return problem.samples
  # End sampler_wrapper

  return sampler_wrapper
# End sampler_decorator()


def analyzer_decorator(func):
  @wraps(func)
  def analyzer_wrapper(*args, **kwargs):

    # I know the below represents some duplication 
    # but it can be cleaned up later
    if 'num_vars' not in problem:
      problem['num_vars'] = len(problem['names'])

    try:
      is_problemspec = issubclass(problem, ProblemSpec)
      has_sampler_info = problem.sampler_name  # None resolves to 'False'
      if not has_sampler_info or not is_problemspec:
        raise AttributeError()
      
    except AttributeError as e:
        warnings.warn("Cannot check for compatible analysis method")
    else:
      # If no exceptions are raised...

      # In this example, `is_compatible` would raise a TypeError if the methods
      # were not compatible. Otherwise it could be made to return `False`
      # and the incompatibility handled here (e.g. raise warning?)
      analyze_name = func.__module__

      try:
        problem.is_compatible(analyze_name)
      except TypeError as e:
        warnings.warn("Sampling and analysis methods are incompatible!\n{}".format(e))


    results = ResultDict(func(*args, **kwargs))
    results.problem = problem
    results.to_df = MethodType(to_df, results)

    return results
  # End analyzer_wrapper

  return analyzer_wrapper
# End analyzer_decorator()
```

Then all that is left to do is decorate the existing functions

```python
# In latin.py
@sampler_decorator
def latin_sample(problem, N, seed=None):
  # the usual latin sampler but is now decorated to return 
  # the custom ProblemSpec class. Behaviour should be identical.


# In sobol.py
@analyzer_decorator
def analyze(...):
  # same as above
```

Are there any thoughts, objections, or things/use cases that I haven't thought about that the above wouldn't address?

**Comment by willu47** [comment]
Hi @ConnectedSystems - many thanks for this. I think a move towards SALib 2.0 is a welcome one, and would provide a good bit of momentum for the package and an opportunity to tie up a number of loose ends.

I think the decorator approach does make the most sense, as it allows us to retain the current package structure but still enforce checking of the problem dict.

Regarding mixing and matching - in some cases, multiple sampling methods can be used with an analyse method (DMIM for example), so that part of the ProblemSpec should be thought through carefully.  Otherwise, I think the ProblemSpec class with sample and analyze methods moves us closer to numpy and pandas approaches, where algorithm methods are specified using names in generic function calls.

**Comment by willu47** [comment]
Another thought is that our emphasis so far has been on providing well tested implementations of common GSA methods. It would be nice to expand the range of methods available, and also make use of the new functionality provided by Python 3.6 onwards, for example, type hints, to further improve confidence in our implementations.

**Comment by ConnectedSystems** [comment]
Re generic function that handles dispatching to specialized methods 

An alternative I was thinking of is instead of a generic method which suggests an `if` block that will grow as more methods are added, 
how about the various methods be handled dynamically. See comparison/example below


```python
# Generic method name that calls relevant function
problem.sample('latin', num_samples=1000)

# vs

# Specific method name tied to the relevant function
problem.sample_latin(num_samples=1000)
```

The advantage of the latter is instead of users having to trawl through an ever expanding docstring which lists out all the 
options/SA methods etc, they can instead use tab-interrogation or similar:

```python
# in Jupyter or in an IDE
problem.sample_[TAB]
-> sample_latin
-> sample_saltelli
-> sample_etc
```

It would be far more discoverable for the user than the usual "go read the manual" approach.

On implementation of SA methods - particularly the analysis methods - are there any changes to the SALib framework that would make development and contribution easier? I don't have anything in mind, seeing as I've yet to implement anything myself, but you might know of any streamlining that could happen.

I could also have a go at doing a direct port of something, perhaps from the R `sensitivity` package and see if there are any pain points that crop up.


Good suggestion regarding leveraging the type hint infrastructure and other features. Might I suggest a roadmap to v2.0 be hashed out with jdherman?

It'll be good to know what features and enhancements have been thought of and their relative desirability and priority.

For example, I personally would put a goal of "all core SA methods have type hinting" closer towards v2.0, e.g. v1.8 or there abouts.

**Comment by jdherman** [comment]
@ConnectedSystems thanks, this is a really thoughtful overview. Sorry for the slow reply - we can definitely work on a roadmap to 2.0.

You've picked out some of the biggest design flaws, especially (3) mixing and matching methods. The original idea was to keep everything decoupled for people whose models aren't written in python ... but this has led to perhaps too much flexibility. I would love to see this problem fixed, among others. 

We'll think it over for a bit and run some ideas by you. More soon. And thanks!!

**Comment by JoerivanEngelen** [comment]
A suggestion: (perhaps for v3) to speed up certain steps (mainly thinking about the brute-forcing Morris sampling), we could consider using Numba https://numba.pydata.org/ to speed up the bottlenecks. 

As I have only used the Morris sampling functions, I do not know if calculation time is a bottleneck for the other sampling methods. If that is not the case, than including this dependency is a bit overkill.

**Comment by ConnectedSystems** [comment]
I was thinking about experimenting with moving some of the core functions to Cython. Performance then would be more consistent compared to Numba (which has a warm-up period due to its JIT nature which would hurt performance for short runs). That said, it won't happen any time soon, if it even comes to it.

I'm sure there's some fairly minor changes that would improve performance by a noticeable margin so I think we should concentrate on that first, at least for the time being.

While we're on this topic, did you (@JoerivanEngelen) get a chance to look at my changes I PR'd to your branch?

**Comment by jdherman** [comment]
Thanks again @ConnectedSystems for starting this discussion. I'm starting to organize the main open issues into categories we can think about for v2.0: https://github.com/SALib/SALib/projects/1

Let's keep this issue as "object-oriented problem class & error checking", I'll label it as a priority under the structure category.

This is a ton of work, much more than the other items, so feel free to break it up into separate issues if that makes sense. For now I think it's helpful to have the full vision in one place.

**Comment by ConnectedSystems** [comment]
Hi everyone,

I have a proof of concept working in a [branch of mine](https://github.com/ConnectedSystems/SALib/tree/problem-spec) if you care to have a look and play around.

The big change is as discussed above although the implementation is slightly different.

Rather than decorators or re-working everything into classes, I implemented a generic interface through which sampling, model evaluation, and analysis will occur. This allows contributors to just provide the key functions for the SA methods, maintains backward compatibility, and can potentially simplify the codebase.

One benefit is that we can provide parallel model runs.

A working example can be found under the [examples directory](https://github.com/ConnectedSystems/SALib/blob/problem-spec/examples/Problem/problem_spec.py) but to give a peek:

```python
from SALib.analyze import sobol
from SALib.sample import saltelli
from SALib.test_functions import Ishigami
from SALib import ProblemSpec

sp = ProblemSpec({
        # `num_vars` key no longer mandatory and is automatically created
        'names': ['x1', 'x2', 'x3'],  
        'groups': None,
        'bounds': [[-3.14159265359, 3.14159265359],
                         [-3.14159265359, 3.14159265359],
                         [-3.14159265359, 3.14159265359]],
        'outputs': ['Y']  # new 'outputs' key required
    })

# Single process execution is simply `run()`
# Yes, this is generic so all existing sampling and analysis functions should work.
#    (although I have not tested these)
# All model run wrappers should work so long as they accept a numpy array of inputs
# as the first parameter
(sp.sample(saltelli.sample, 200000, calc_second_order=True)
        .run_parallel(Ishigami.evaluate)  
        .analyze(sobol.analyze, calc_second_order=True, conf_level=0.95))

print(sp)
```

The final print statement there outputs:

```bash
Samples: (1600000, 3) 

Outputs: (1600000,) 

Analysis:

          ST   ST_conf
x1  0.557589  0.006344
x2  0.442408  0.002778
x3  0.243672  0.001878
          S1   S1_conf
x1  0.313672  0.004408
x2  0.442430  0.003741
x3 -0.000261  0.003955
                S2   S2_conf
(x1, x2)  0.000039  0.005633
(x1, x3)  0.243950  0.007465
(x2, x3)  0.000011  0.004694
```

Note the very high number of samples used - that's how many you need to show some benefit in parallelizing the Ishigami runs. Otherwise single process is faster simply due to the overheads associated with setting up parallel processes and moving data to and fro, so this should be reserved for CPU intensive models.

If we change the Ishigami implementation to vectorize the calculations, then multiprocessing becomes essentially pointless, at least with how I've currently implemented it. That said, by no means do I claim the parallel processing is as efficient as it could be, but it's working.

```python
def evaluate(values):
    """Vectorized Ishigami."""
    A = 7
    B = 0.1

    X = values
    Y = np.sin(X[:, 0]) + A * np.power(np.sin(X[:, 1]), 2) + \
            B * np.power(X[:, 2], 4) * np.sin(X[:, 0])
    
    return Y
```

As I said, feel free to play around with it and give me feedback.

**Comment by jdherman** [comment]
@ConnectedSystems , thanks, this is a big step for the project roadmap. This seems like the perfect compromise between backward compatibility and adding enough OO structure to be able to clean up the redundant and messy error checking within each method. 

I'll keep thinking about it, and curious to get input from others. But maybe it's time to finally start a development branch for a new version. We had initially planned for this to be part of v2.0, but since your solution is cleaner and won't break anything, it could also target something sooner (1.5?)

**Comment by ConnectedSystems** [comment]
I like your thinking @jdherman . 

I realised recently that we hadn't explicitly discussed what the migration plan would be for users of SALib. I've assumed that we expect all users to adopt this approach (whatever "this approach" ends up being) and leave backward compatibility only for legacy reasons, and so support will eventually be phased out.

In any case I agree it would be a good idea to introduce this approach earlier than v2.0 proper. I'd like deprecation warnings to be thrown, and such warnings could direct users to information on how to upgrade their workflow. We'd also need to update the docs of course.

With all that, perhaps v1.5 is ambitious? I'm not completely aware of the time frame but v1.5 sounds relatively close. I'm unsure if I/we can get all the necessary pieces done and don't want to place undue pressure on ourselves.

On a personal level, I'm nearing the end of my PhD so much of my time is on finishing up papers and the thesis itself. I am doing it by publication so the compilation of the text is less arduous than it otherwise would be, but doesn't (shouldn't?) mean I can spend a lot of time on this in the near-term.

I'd also like to see issue #271 resolved before further releases are made.

**Comment by jdherman** [comment]
Yea when I put together the roadmap there wasn't really a time frame associated with it. That's all up for discussion. New features are always great, but can't compete with finishing your PhD (congrats!)

**Comment by ConnectedSystems** [comment]
A small update - allowing interactive discoverability of available sampling/analysis methods as I previously envisioned.

The feature is implemented dynamically, by scanning the relevant "sample" or "analyze" modules and importing the relevant function (i.e. `sample()` or `analyze()`).

![image](https://user-images.githubusercontent.com/4075136/83355318-2dfff500-a3a2-11ea-987c-8ec0ce030bfa.png)

![image](https://user-images.githubusercontent.com/4075136/83355333-3e17d480-a3a2-11ea-9ac4-49ffa9d1858a.png)


The `sample()` and `analyze()` are kept as generic implementations (usage shown in my earlier comment [above](https://github.com/SALib/SALib/issues/234#issuecomment-623444497)) which allows users to define their own functions and, theoretically lets them use these custom implementations in conjunction with anything SALib offers.

**Comment by ConnectedSystems** [comment]
Did my final (exit) seminar for the PhD this morning so in celebration I did a little bit more work on this.

All sampling and analysis methods now get dynamically wrapped attached to the SALib ProblemSpec object. It now handles automatic passing of both X and Y values if the analysis method needs it - the caveat is that it currently only scans for the `X` function argument.

All this means you can now do something like this:

```python
(sp.sample_saltelli(100000, calc_second_order=True)
    .run_parallel(Ishigami.evaluate)  
    .analyze_sobol(calc_second_order=True, conf_level=0.95))

print(sp)
```

Which outputs:

```bash
100%███████████████████████████████████████████████████| 8/8 [00:07<00:00,  1.08it/s]

Samples: (800000, 3) 

Outputs: (800000,)

Analysis:

          ST   ST_conf
x1  0.557686  0.008672
x2  0.442305  0.003910
x3  0.243704  0.002617

          S1   S1_conf
x1  0.313681  0.006069
x2  0.442440  0.005114
x3 -0.000385  0.005845

                S2   S2_conf
(x1, x2) -0.000195  0.008975
(x1, x3)  0.244033  0.011149
(x2, x3) -0.000100  0.005907
```

The original docstring gets retained as before - I may look into modding the function arg list so that unnecessary details get removed. For example the problem spec and X/Y values get automatically passed in (as mentioned earlier) if using the ProblemSpec class.

**Comment by ConnectedSystems** [comment]
This is now included in v1.4, thanks everyone for the discussion input.
